app/controllers/templates_controller.rb

```diff
@@ -0,0 +1,147 @@
+class TemplatesController < ApplicationController
+  include Foreman::Renderer
+  include Foreman::Controller::ProvisioningTemplates
+  include Foreman::Controller::AutoCompleteSearch
+
+  before_filter :handle_template_upload, :only => [:create, :update]
+  before_filter :find_resource, :only => [:edit, :update, :destroy, :clone, :lock, :unlock]
+  before_filter :load_history, :only => :edit
+  before_filter :type_name_plural, :type_name_singular, :resource_class
+
+  include TemplatePathsHelper
+
+  def index
+    @templates = resource_base.search_for(params[:search], :order => params[:order]).paginate(:page => params[:page])
+    @templates = @templates.includes(resource_base.template_includes)
+    render template_view('index')
+  end
+
+  def new
+    @template = resource_class.new
+    render template_view('new')
+  end
+
+  def clone
+    @template = @template.dup
+    load_vars_from_template
+    flash[:warning] = _("The marked fields will need reviewing")
+    @template.valid?
+    render template_view('new')
+  end
+
+  def lock
+    set_locked true
+  end
+
+  def unlock
+    set_locked false
+  end
+
+  def create
+    @template = resource_class.new(params[type_name_singular])
+    if @template.save
+      process_success :success_redirect => template_path_for(params[:type])
+    else
+      process_error :render => template_view('new')
+    end
+  end
+
+  def edit
+    load_vars_from_template
+    render template_view('edit')
+  end
+
+  def update
+    if @template.update_attributes(params[type_name_singular])
+      process_success :success_redirect => template_path_for(params[:type])
+    else
+      load_history
+      process_error :render => template_view('edit')
+    end
+  end
+
+  def revision
+    audit = Audit.find(params[:version])
+    render :json => audit.revision.template
+  end
+
+  def destroy
+    if @template.destroy
+      process_success :success_redirect => template_path_for(params[:type])
+    else
+      process_error :redirect => template_path_for(params[:type])
+    end
+  end
+
+  private
+
+  def template_view(name)
+    "#{@type_name_plural}/#{name}"
```

Comment:
I would recommend using `type_name_plural` instead of `@type_name_plural` to prevent issues, when the variable is not initialized yet